File: main_project/tripod-new/src/cpp/internalclient/CacheManagerClient.cpp
#include <CacheManagerClient.h>
#include "boost/foreach.hpp"
#include <Common.h>
#include <CacheManagerConfigClient.h>


namespace com {
namespace renren {
namespace tripod {

CacheManagerClient::CacheManagerClient(const std::string& zkAddress) : zkAddress_(zkAddress) {
}

CacheManagerConfigClientPtr CacheManagerClient::getConfig() {
    if(config_ == NULL) {
        boost::upgrade_lock<boost::shared_mutex> lock(configMutex_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        config_ = ConfigClientFactory::getInstance()->getCacheManagerConfigClient(zkAddress_);
    }
    return config_;
}


DataMap CacheManagerClient::get(const KeySeq& keys, const std::string& namespaceId, 
        const std::string& businessId, int expiredTime, bool useMaster, const long timeout) {
    DataMap res;
   
    CacheManagerConfigClientPtr config = getConfig();
    if(config == NULL) {
        TRIPOD_WARN("CacheManagerClient::get config is NULL, namespaceId " + namespaceId + " businessId " + businessId);
        return res;
    }
    
    std::map<CacheManagerPrx, KeySeq> prxMap = config->getPrxToKeysMap(keys);

    for(std::map<CacheManagerPrx, KeySeq>::const_iterator it = prxMap.begin(); it != prxMap.end(); ++it) {
        try {
            DataMap singleRes = it->first->ice_twoway()->ice_timeout(timeout)->get(it->second, namespaceId, businessId, expiredTime, useMaster);
            res.insert(singleRes.begin(), singleRes.end());
        }catch(Ice::Exception& e) {
            std::ostringstream os;
            os << "CacheManagerClient::get Ice::Exception " << e.what() << " namespaceId " << namespaceId << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }catch(...) {
            std::ostringstream os;
            os << "CacheManagerClient::get UnKnownException namespaceId " << namespaceId << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }
    return res;
}

bool CacheManagerClient::load(const KeySeq& keys, const std::string& namespaceId, 
        const std::string& businessId, int expiredTime, bool useMaster, const long timeout) {
   
    CacheManagerConfigClientPtr config = getConfig();
    if(config == NULL) {
        TRIPOD_WARN("CacheManagerClient::load config is NULL, namespaceId " + namespaceId + " businessId " + businessId);
        return false;
    }
    
    std::map<CacheManagerPrx, KeySeq> prxMap = config->getPrxToKeysMap(keys);

    bool flag = true;
    bool inFlag = false;

    for(std::map<CacheManagerPrx, KeySeq>::const_iterator it = prxMap.begin(); it != prxMap.end(); ++it) {
        inFlag = true;
        try {
            flag &= it->first->ice_twoway()->ice_timeout(timeout)->load(it->second, namespaceId, businessId, expiredTime, useMaster);
        }catch(Ice::Exception& e) {
            std::ostringstream os;
            os << "CacheManagerClient::load Ice::Exception " << e.what() << " namespaceId " << namespaceId << " businessId " << businessId;
            os << " timeout: " << timeout << " keys.size: " << keys.size();
            if (keys.size() < 1000) {
              os << " keys:";
              for (KeySeq::const_iterator it = keys.begin(); it != keys.end(); ++it) {
                os << *it << ",";
              }
            }
            TRIPOD_WARN(os.str());
        }catch(...) {
            std::ostringstream os;
            os << "CacheManagerClient::load UnKnownException namespaceId " << namespaceId << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }
    return flag && inFlag;
}

bool CacheManagerClient::deleteCache(const KeySeq& keys, const std::string& namespaceId, 
        const std::string& businessId, const long timeout) {

    CacheManagerConfigClientPtr config = getConfig();
    if(config == NULL) {
        return false;
    }
    
    std::map<CacheManagerPrx, KeySeq> prxMap = config->getPrxToKeysMap(keys);

    bool flag = true;
    bool inFlag = false;

    for(std::map<CacheManagerPrx, KeySeq>::const_iterator it = prxMap.begin(); it != prxMap.end(); ++it) {
        inFlag = true;
        try {
            flag &= it->first->ice_twoway()->ice_timeout(timeout)->deleteCache(it->second, namespaceId, businessId);
        }catch(Ice::Exception& e) {
            std::ostringstream os;
            os << "CacheManagerClient::deleteCache Ice::Exception " << e.what() << " namespaceId " << namespaceId << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }catch(...) {
            std::ostringstream os;
            os << "CacheManagerClient::deleteCache UnKnownException namespaceId " << namespaceId << " businessId " << businessId;
            TRIPOD_WARN(os.str());
        }
    }
    return flag && inFlag;
}


bool CacheManagerClient::LoadListCache(const std::string& key, const std::string& namespace_id, 
                                                const std::string& biz_id, bool use_master, 
                                                const long timeout) {
  CacheManagerConfigClientPtr config = getConfig();
  if(!config) {
    MCE_WARN("CacheManagerClient::LoadListCache config is NULL, namespaceId " + namespace_id + " businessId " + biz_id);
    return false;
  }

  KeySeq keys;
  keys.push_back(key);

  bool flag = true;

  std::map<CacheManagerPrx, KeySeq> prx_map = config->getPrxToKeysMap(keys);
  std::map<CacheManagerPrx, KeySeq>::const_iterator iter = prx_map.begin();
  for (; iter != prx_map.end(); ++iter) { // just for one loop
    const KeySeq& seq = iter->second;
    BOOST_FOREACH(const std::string& k, seq) {
      try {
        iter->first->ice_twoway()->ice_timeout(timeout)->loadListCache(k, namespace_id, biz_id, use_master);
      } catch (Ice::Exception& e) {
        MCE_WARN("CacheManagerConfigClientPtr::LoadListCache() " << e.what() << "namespace_id:" << namespace_id << " business_id:" << biz_id);
        flag = false;
      } catch (...) {
        MCE_WARN("CacheManagerConfigClientPtr::LoadListCache() unknown exception ...  namespace_id:" << namespace_id << " business_id:" << biz_id);
        flag = false;
      }
    }
  }
  return flag;
}

bool CacheManagerClient::LoadHashCache(const std::string& key, const std::string& namespace_id, 
                                                const std::string& biz_id, bool use_master, 
                                                const long timeout) {
  CacheManagerConfigClientPtr config = getConfig();
  if(!config) {
    MCE_WARN("CacheManagerClient::LoadHashCache config is NULL, namespaceId " + namespace_id + " businessId " + biz_id);
    return false;
  }

  KeySeq keys;
  keys.push_back(key);
  bool flag = true;
  std::map<CacheManagerPrx, KeySeq> prx_map = config->getPrxToKeysMap(keys);
  std::map<CacheManagerPrx, KeySeq>::const_iterator iter = prx_map.begin();
  for (; iter != prx_map.end(); ++iter) { // just for one loop
    const KeySeq& seq = iter->second;
    BOOST_FOREACH(const std::string& k, seq) {
      try {
        iter->first->ice_twoway()->ice_timeout(timeout)->loadHashCache(k, namespace_id, biz_id, use_master);
      } catch (Ice::Exception& e) {
        MCE_WARN("CacheManagerConfigClientPtr::LoadHashCache() " << e.what() << "namespace_id:" << namespace_id << " business_id:" << biz_id);
        flag = false;
      } catch (...) {
        MCE_WARN("CacheManagerConfigClientPtr::LoadHashCache() unknown exception ...  namespace_id:" << namespace_id << " business_id:" << biz_id);
        flag = false;
      }
    }
  }
  return flag;
}


bool CacheManagerClient::DeleteListOrHashCache(const std::string& key, const std::string& namespace_id, 
                                               const std::string& biz_id, const long timeout) {
  CacheManagerConfigClientPtr config = getConfig();
  if(!config) {
    MCE_WARN("CacheManagerClient::DeleteListOrHashCache config is NULL, namespaceId " + namespace_id + " businessId " + biz_id);
    return false;
  }

  KeySeq keys;
  keys.push_back(key);
  bool flag = true;

  std::map<CacheManagerPrx, KeySeq> prx_map = config->getPrxToKeysMap(keys);
  std::map<CacheManagerPrx, KeySeq>::const_iterator iter = prx_map.begin();
  for (; iter != prx_map.end(); ++iter) {
    const KeySeq& seq = iter->second;
    BOOST_FOREACH(const std::string& k, seq) {
      try {
        iter->first->ice_twoway()->ice_timeout(timeout)->deleteListOrHashCache(k, namespace_id, biz_id);
      } catch (Ice::Exception& e) {
        MCE_WARN("CacheManagerConfigClientPtr::DeleteListOrHashCache() " << e.what() << "namespace_id:" << namespace_id << " business_id:" << biz_id);
        flag = false;
      } catch (...) {
        MCE_WARN("CacheManagerConfigClientPtr::DeleteListOrHashCache() unknown exception ...  namespace_id:" << namespace_id << " business_id:" << biz_id);
        flag = false;
      }
    }
  }
  return flag;
}

BigKeyDataMap CacheManagerClient::GetHashCache(const std::string& key, const StrList& fields, const std::string& namespace_id, 
                                                const std::string& biz_id, bool use_master, const long timeout) {
  BigKeyDataMap ret_data;
  CacheManagerConfigClientPtr config = getConfig();
  if(!config) {
    MCE_WARN("CacheManagerClient::GetHashCache config is NULL, namespaceId " + namespace_id + " businessId " + biz_id);
    return ret_data;
  }
  KeySeq keys;
  keys.push_back(key);
  std::map<CacheManagerPrx, KeySeq> prx_map = config->getPrxToKeysMap(keys);
  std::map<CacheManagerPrx, KeySeq>::const_iterator iter = prx_map.begin();
  for (; iter != prx_map.end(); ++iter) { // just for one loop
    const KeySeq& seq = iter->second;
    BOOST_FOREACH(const std::string& k, seq) {
      try {
        ret_data =  iter->first->ice_twoway()->ice_timeout(timeout)->getHashCache(k, fields, namespace_id, biz_id, use_master);
      } catch (Ice::Exception& e) {
        MCE_WARN("CacheManagerConfigClientPtr::GetHashCache() " << e.what() << "namespace_id:" << namespace_id << " business_id:" << biz_id);
      } catch (...) {
        MCE_WARN("CacheManagerConfigClientPtr::GetHashCache() unknown exception ...  namespace_id:" << namespace_id << " business_id:" << biz_id);
      }
    }
  }
  return ret_data;
}

}}} // com::renren::tripod

File: main_project/tripod-new/src/java/TripodInternalClient/src/main/java/com/renren/tripod/internalclient/CacheManagerClient.java
package com.renren.tripod.internalclient;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import com.renren.tripod.CacheManagerPrx;
import com.renren.tripod.CacheManagerPrxHelper;
import com.renren.tripod.configclient.CacheManagerConfigClient;
import com.renren.tripod.configclient.ConfigClientFactory;
import com.xiaonei.xce.log.Tripod;

public class CacheManagerClient {

    private CacheManagerConfigClient _config = null;
    private String _zkAddress = null;

    public CacheManagerClient(String zkAddress) {
	_zkAddress = zkAddress;
    }

    private CacheManagerConfigClient getConfigClient() {
	if (_config == null) {
	    synchronized (this) {
		_config = ConfigClientFactory
			.getCacheMangerConfigClient(_zkAddress);
	    }
	}
	return _config;
    }

    public Map<String, byte[]> get(List<String> keys, String namespaceId,
	    String businessId, int expiredTime, boolean useMaster) {
	return get(keys, namespaceId, businessId, expiredTime, useMaster, 300);
    }

    public Map<String, byte[]> get(List<String> keys, String namespaceId,
	    String businessId, int expiredTime, boolean useMaster, int timeout) {
	CacheManagerConfigClient config = getConfigClient();
	if (config == null) {
	    Tripod.getLogger().error(
		    "CacheManagerClient.get config == null, namespaceId "
			    + namespaceId + " businessId" + businessId);
	    return null;
	}
	Map<CacheManagerPrx, List<String>> prxMap = config
		.getPrxToKeysMap(keys);
	Map<String, byte[]> res = new HashMap<String, byte[]>();
	for (Entry<CacheManagerPrx, List<String>> ent : prxMap.entrySet()) {
	    CacheManagerPrx prx = CacheManagerPrxHelper.uncheckedCast(ent
		    .getKey().ice_twoway().ice_timeout(timeout));
	    Map<String, byte[]> singleRes = prx.get(
		    keys.toArray(new String[keys.size()]), namespaceId,
		    businessId, expiredTime, useMaster);
	    res.putAll(singleRes);
	}
	return res;
    }

    public boolean load(List<String> keys, String namespaceId,
	    String businessId, int expiredTime, boolean useMaster) {
	return load(keys, namespaceId, businessId, expiredTime, useMaster, 1000);
    }

    public boolean load(List<String> keys, String namespaceId,
	    String businessId, int expiredTime, boolean useMaster, int timeout) {
	CacheManagerConfigClient config = getConfigClient();
	if (config == null) {
	    Tripod.getLogger().error(
		    "CacheManagerClient.load config == null, namespaceId "
			    + namespaceId + " businessId " + businessId);
	    return false;
	}
	Map<CacheManagerPrx, List<String>> prxMap = config
		.getPrxToKeysMap(keys);
	boolean sucFlag = true;
	for (Entry<CacheManagerPrx, List<String>> ent : prxMap.entrySet()) {
	    CacheManagerPrx prx = CacheManagerPrxHelper.uncheckedCast(ent
		    .getKey().ice_twoway().ice_timeout(timeout));
	    boolean flag = prx.load(
		    ent.getValue().toArray(new String[ent.getValue().size()]),
		    namespaceId, businessId, expiredTime, useMaster);
	    sucFlag &= flag;
	}
	return sucFlag;
    }

    public boolean deleteCache(List<String> keys, String namespaceId,
	    String businessId) {
	return deleteCache(keys, namespaceId, businessId, 1000);
    }

    public boolean deleteCache(List<String> keys, String namespaceId,
	    String businessId, int timeout) {
	CacheManagerConfigClient config = getConfigClient();
	if (config == null) {
	    Tripod.getLogger().error(
		    "CacheManagerClient.deleteCache config == null, namespaceId "
			    + namespaceId + " businessId " + businessId);
	    return false;
	}
	Map<CacheManagerPrx, List<String>> prxMap = config
		.getPrxToKeysMap(keys);
	boolean sucFlag = true;
	for (Entry<CacheManagerPrx, List<String>> ent : prxMap.entrySet()) {
	    CacheManagerPrx prx = CacheManagerPrxHelper.uncheckedCast(ent
		    .getKey().ice_twoway().ice_timeout(timeout));
	    boolean flag = prx.deleteCache(
		    keys.toArray(new String[keys.size()]), namespaceId,
		    businessId);
	    sucFlag &= flag;
	}
	return sucFlag;
    }

    public boolean loadHashCache(String key, String namespaceId, String bizId,
	    boolean useMaster, int timeout) {
	boolean result = false;
	CacheManagerConfigClient config = getConfigClient();
	if (config == null) {
	    Tripod.getLogger().error(
		    "CacheManagerClient.loadHashCache config == null, namespaceId "
			    + namespaceId + " businessId " + bizId);
	    return false;
	}
	List<String> keys = new ArrayList<String>();
	keys.add(key);
	Map<CacheManagerPrx, List<String>> proxyMap = config
		.getPrxToKeysMap(keys);
	for (Entry<CacheManagerPrx, List<String>> entry : proxyMap.entrySet()) {
	    try {
		CacheManagerPrx proxy = CacheManagerPrxHelper
			.uncheckedCast(entry.getKey().ice_twoway()
				.ice_timeout(timeout));
		for (String k : entry.getValue()) { // just one loop
		    result = proxy.loadHashCache(k, namespaceId, bizId,
			    useMaster);
		}
	    } catch (Exception e) {
		e.printStackTrace();
		result = false;
		break;
	    }

	}
	return result;
    }

    public boolean deleteListOrHashCache(String key, String namespaceId,
	    String bizId, int timeout) {
	boolean result = false;
	CacheManagerConfigClient config = getConfigClient();
	if (config == null) {
	    Tripod.getLogger().error(
		    "CacheManagerClient.deleteListOrHashCache config == null, namespaceId "
			    + namespaceId + " businessId " + bizId);
	    return false;
	}
	List<String> keys = new ArrayList<String>();
	keys.add(key);
	Map<CacheManagerPrx, List<String>> proxyMap = config
		.getPrxToKeysMap(keys);
	for (Entry<CacheManagerPrx, List<String>> entry : proxyMap.entrySet()) {
	    try {
		CacheManagerPrx proxy = CacheManagerPrxHelper
			.uncheckedCast(entry.getKey().ice_twoway()
				.ice_timeout(timeout));
		for (String k : entry.getValue()) { // just one loop
		    result = proxy.deleteListOrHashCache(k, namespaceId, bizId);
		}
	    } catch (Exception e) {
		e.printStackTrace();
		result = false;
		break;
	    }

	}
	return result;
    }

    public Map<String, String> getHashCache(String key, List<String> fields,
	    String namespaceId, String bizId, boolean useMaster, int timeout) {
	Map<String, String> result = new HashMap<String, String>();
	CacheManagerConfigClient config = getConfigClient();
	if (config == null) {
	    Tripod.getLogger().error(
		    "CacheManagerClient.getHashCache config == null, namespaceId "
			    + namespaceId + " businessId " + bizId);
	    return result;
	}
	List<String> keys = new ArrayList<String>();
	keys.add(key);
	Map<CacheManagerPrx, List<String>> proxyMap = config
		.getPrxToKeysMap(keys);
	for (Entry<CacheManagerPrx, List<String>> entry : proxyMap.entrySet()) {
	    try {
		CacheManagerPrx proxy = CacheManagerPrxHelper
			.uncheckedCast(entry.getKey().ice_twoway()
				.ice_timeout(timeout));
		for (String k : entry.getValue()) { // just one loop
		    String[] fieldArray = new String[fields.size()];
		    for (int i = 0; i < fields.size(); ++i) {
			fieldArray[i] = fields.get(i);
		    }
		    result = proxy.getHashCache(key, fieldArray, namespaceId,
			    bizId, useMaster);
		}
	    } catch (Exception e) {
		e.printStackTrace();
		break;
	    }
	}
	return result;
    }
}
